linera-base/src/tracing.rs

```diff
@@ -3,79 +3,130 @@
 
 //! This module provides unified handling for tracing subscribers within Linera binaries.
 
-/// Initializes tracing for the browser, sending messages to the developer console and
-/// span events to the [Performance
-/// API](https://developer.mozilla.org/en-US/docs/Web/API/Performance).
-#[cfg(web)]
-pub fn init() {
-    use tracing_subscriber::{
-        prelude::__tracing_subscriber_SubscriberExt as _, util::SubscriberInitExt as _,
-    };
+use std::{
+    env,
+    fs::{File, OpenOptions},
+    path::Path,
+    sync::Arc,
+};
 
-    tracing_subscriber::registry()
-        .with(
-            tracing_subscriber::fmt::layer()
-                .with_ansi(false)
-                .without_time()
-                .with_writer(tracing_web::MakeWebConsoleWriter::new()),
-        )
-        .with(
-            tracing_web::performance_layer()
-                .with_details_from_fields(tracing_subscriber::fmt::format::Pretty::default()),
-        )
-        .init();
-}
+use is_terminal::IsTerminal as _;
+use tracing::Subscriber;
+use tracing_subscriber::{
+    fmt::{
+        self,
+        format::{FmtSpan, Format, Full},
+        time::FormatTime,
+        FormatFields, MakeWriter,
+    },
+    layer::{Layer, SubscriberExt as _},
+    registry::LookupSpan,
+    util::SubscriberInitExt,
+};
 
 /// Initializes tracing in a standard way.
+///
 /// The environment variables `RUST_LOG`, `RUST_LOG_SPAN_EVENTS`, and `RUST_LOG_FORMAT`
 /// can be used to control the verbosity, the span event verbosity, and the output format,
 /// respectively.
-#[cfg(not(web))]
-pub fn init() {
-    use is_terminal::IsTerminal as _;
-    use tracing_subscriber::fmt::format::FmtSpan;
-
-    fn fmt_span_from_str(events: &str) -> FmtSpan {
-        let mut fmt_span = FmtSpan::NONE;
-        for event in events.split(',') {
-            fmt_span |= match event {
-                "new" => FmtSpan::NEW,
-                "enter" => FmtSpan::ENTER,
-                "exit" => FmtSpan::EXIT,
-                "close" => FmtSpan::CLOSE,
-                "active" => FmtSpan::ACTIVE,
-                "full" => FmtSpan::FULL,
-                _ => FmtSpan::NONE,
-            };
-        }
-        fmt_span
-    }
+///
+/// The `LINERA_LOG_DIR` environment variable can be used to configure a directory to
+/// store log files. If it is set, a file with the `log_file_name` is created in the
+/// directory.
+pub fn init(log_file_name: impl AsRef<Path>) {
```

Comment:
Can we pass a "process name" instead? (Concretely, remove the `.log` below)
In this way, we can enrich the `stderr` output even we don't dump to files.

Reply:
Incidentally, we don't know that the path is relative here.

Reply:
I'm not sure I understand. You mean rename the parameter, change it's type to be just a string (`&str`?), and automatically append `.log` to it when building the `log_file_path`?

Reply:
Yes that's right

Reply:
Done.